### DISCUSSION TITLE
How to determine whether the parameters of the method are allowed to be empty?

### DISCUSSION THREAD

**Discussion by NMSAzulX** [dotnet/runtime#61499] [Q&A]
e.g.
```c#

public class Test{

public void Show1(string name){}

public void Show2(string? name){}
}
```

I did some experiments and found that PatameterInfo has no attribute to express that the parameter can be empty.
Is there any good idea?

**Comment by Clockwork-Muse** [comment]
If you mean "express that the parameter can be **null** ", then it's [apparently done as an attribute](https://stackoverflow.com/a/58454489/812837).

**Reply by svick** [reply]
And if you look at the answer below that, you'll learn that [.Net 6 has an API for that](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.nullabilityinfocontext?WT.mc_id=DOP-MVP-5002735).

**Reply by NMSAzulX** [reply]
Thank you! It works. Can this method detect string?[] is there a null reference?  GenericType or ElementType.

**Reply by NMSAzulX** [reply]
> And if you look at the answer below that, you'll learn that [.Net 6 has an API for that](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.nullabilityinfocontext?WT.mc_id=DOP-MVP-5002735).

This API is really easy to use. But I'm working on netcore3.1 and 5. 0 compatibility, so I need another way. Do you have any suggestions?

**Reply by huoyaoyuan** [reply]
There isn't OOB package shipped, so you have to port the code manually.

**Comment by NMSAzulX** [comment]
Now, I have to port the source code and comment out some internal code. I'm not sure what problems this will cause. The three classes migrated to 3.1 or 5.0 are: `NullabilityInfo` ,`NullabilityInfoContext`, `NullabilityState`  

Comment out some internal code
```c#
 private static MemberInfo GetMemberMetadataDefinition(MemberInfo member)
 {
        //Type? type = member.DeclaringType;
        //if ((type != null) && type.IsGenericType && !type.IsGenericTypeDefinition)
        //{
        //    return type.GetGenericTypeDefinition().GetMemberWithSameMetadataDefinitionAs(member);
       //}
       return member;
}
```

**Comment by Xyncgas** [comment]
The IDE highlights stuffs.
For example if you are assigning null to T instead of T?
But if you ignore it when IDE warns you or if you are using command lines to compile then it's going to compile as of 2021.

**Reply by Xyncgas** [reply]
Unless you use if check you can't ensure it's null.
maybe I am missing something here though, because I heard about System.Nullable<T> and how you can use HasValue to check

**Reply by Clockwork-Muse** [reply]
In context, OP appears to be using reflection to extract information about methods - there's not going to be an IDE involved, and since the method itself is not being executed there's no way to check `HasValue` either.

**Reply by NMSAzulX** [reply]
Yes. Now I'm going to pause some work on nullable reference reflection because it's hard to reflect, such as method's return values, passing in generics, and so on.
